source/common/upstream/original_dst_cluster.h

```diff
@@ -0,0 +1,95 @@
+#pragma once
+
+#include <cstdint>
+#include <functional>
+#include <string>
+
+#include "envoy/thread_local/thread_local.h"
+
+#include "common/common/empty_string.h"
+#include "common/common/logger.h"
+#include "common/upstream/upstream_impl.h"
+
+namespace Envoy {
+namespace Upstream {
+
+/**
+ * The OriginalDstCluster is a type of cluster that creates a single
+ * logical host that allows upstream connections to be opened to
+ * arbitrary IP destinations, as directed by metadata made available
+ * earlier in the filter pipeline.  Connection requests to the logical
+ * host itself will be denied.
+ *
+ * A new 'real host' is created for each destination, but currently
+ * stats and other state refers to the logical host.
+ */
+class OriginalDstCluster : public ClusterImplBase {
```

Comment:
The major problem I see with this approach (which is not quite the same as LogicalDnsCluster) is that you have the potential to add an unlimited # of hosts (and by association connection pools) that are never going to get cleaned up once they are created. At minimum I think we would need stats specific to this cluster type telling us how many real hosts we have. We don't specifically track the # of connection pools in stats for HTTP stuff, so it's going to be hard to tell exactly how many hosts/conn_pools we have created.

With that said, you can of course run out of memory in any number of other ways, and presumably you are using this cluster type in a specific deployment where you probably have control over what is happening, but I still feel that there should be some type of output telling us what is going on, even if the hosts aren't specifically in the primary cluster. 

Note also that the way this is implemented your "real hosts" can/will be duplicated across all worker threads. I'm not sure if this matters or not.

Stepping back a bit, there is another approach that could be taken (just sketching it out a bit):

- Write a filter that would sit inline to check to see if the original_dst is already in your cluster
- If not, pause iteration, and do worker <-> main thread sync to add it to cluster, then continue iteration.
- If it is, just continue, and do normal LB with some type of hint that picks a specific host in the load balancer by searching.

I haven't thought through the previous very fully, but if that interests you, we can discuss further (and maybe should do a quick video conference).

I guess given how self contained this change is (mostly) I don't have a huge objection to the approach as long as the downsides are understood and well documented. LMK what you think.